osquery/core/scope_guard.h

```diff
@@ -0,0 +1,58 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed as defined on the LICENSE file found in the
+ *  root directory of this source tree.
+ */
+
+#pragma once
+
+#include <utility>
+
+namespace osquery {
+
+/**
+ * The RAII based scope guard class.
+ *
+ * To be sure that resources are always released/removed/closed/verified/stoped
+ * in face of multiple return statements from the function.
+ *
+ * It takes functor object by value during the construction. It is going to be
+ * called once and only once during the destruction of ScopeGuard object.
+ *
+ * There is helper function to create the object of guard.
+ * @code{.cpp}
+ *   {
+ *     auto const manager = ScopeGuard<>::create(
+ *       [&file_path]() { fs::remove(file_path); }
+ *     );
+ *       ...
+ *     // it will be removed at the end of scope
+ *   }
+ * @endcode
+ */
+template <typename...>
+class ScopeGuard final {
+ public:
+  template <typename FinalRoutineType>
+  static inline auto create(FinalRoutineType final_routine) {
+    return ScopeGuard<FinalRoutineType>(std::move(final_routine));
```

Comment:
Consider using perfect forwarding here to create less copies for non-movable functors

```
template <typename FinalRoutineType>
static inline auto create(FinalRoutineType&& final_routine) {
    return ScopeGuard<FinalRoutineType>(std::forward<FinalRoutineType>(final_routine));
}
```

Note that you would also have to update `ScopeGuard` constructor, although you would not be able to use `std::forward` there, two constructors will have to be implemented